src/Microsoft.AspNetCore.Blazor/RenderTree/RenderTreeBuilder.cs

```diff
@@ -102,6 +102,26 @@ public void AddContent(int sequence, RenderFragment fragment)
             }
         }
 
+        /// <summary>
+        /// Appends frames representing an arbitrary fragment of content.
+        /// </summary>
+        /// <param name="sequence">An integer that represents the position of the instruction in the source code.</param>
+        /// <param name="fragment">Content to append.</param>
+        /// <param name="value">The value used by <paramref name="fragment"/>.</param>
+        public void AddContent<T>(int sequence, RenderFragment<T> fragment, T value)
+        {
+            if (fragment != null)
+            {
+                // We surround the fragment with a region delimiter to indicate that the
+                // sequence numbers inside the fragment are unrelated to the sequence numbers
+                // outside it. If we didn't do this, the diffing logic might produce inefficient
+                // diffs depending on how the sequence numbers compared.
+                OpenRegion(sequence);
+                fragment(this, value);
+                CloseRegion();
```

Comment:
@SteveSandersonMS - intended for use from a component written in C#. Useful?

Reply:
Not sure whether people will use it, but it's totally fine to have it :)

Reply:
I think we'll have to decide which of these two is the core experience, and whether we can avoid allocations in many cases. 

There's feedback elsewhere about a different signature for `RenderFragment<>` that would result in an allocation on each use, but has some nice experience improvments.